File: app/src/main/java/br/com/menuvem/lojista/presentation/insumos/InsumosScreen.kt
package br.com.menuvem.lojista.presentation.insumos

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.Delete
import androidx.compose.material.icons.outlined.Inventory2
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import br.com.menuvem.lojista.domain.model.CategoriaInsumo
import br.com.menuvem.lojista.domain.model.Insumo
import br.com.menuvem.lojista.presentation.components.EmptyState
import br.com.menuvem.lojista.presentation.components.formatarMoeda
import br.com.menuvem.lojista.ui.theme.*

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun InsumosScreen(
    onBack: () -> Unit,
    viewModel: InsumosViewModel = hiltViewModel()
) {
    val uiState by viewModel.uiState.collectAsStateWithLifecycle()
    val snackbarHostState = remember { SnackbarHostState() }

    LaunchedEffect(uiState.error) {
        uiState.error?.let {
            snackbarHostState.showSnackbar(it)
            viewModel.clearError()
        }
    }

    Scaffold(
        snackbarHost = { SnackbarHost(snackbarHostState) },
        topBar = {
            TopAppBar(
                title = { Text("Biblioteca de Insumos") },
                navigationIcon = {
                    IconButton(onClick = onBack) {
                        Icon(
                            imageVector = Icons.AutoMirrored.Filled.ArrowBack,
                            contentDescription = "Voltar"
                        )
                    }
                }
            )
        },
        floatingActionButton = {
            ExtendedFloatingActionButton(
                onClick = { viewModel.onShowForm() },
                icon = { Icon(Icons.Default.Add, contentDescription = null) },
                text = { Text("Novo Insumo") },
                containerColor = YellowSecondary,
                contentColor = MaterialTheme.colorScheme.surface
            )
        }
    ) { paddingValues ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(paddingValues)
        ) {
            Row(
                modifier = Modifier.padding(horizontal = 16.dp, vertical = 4.dp),
                horizontalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                FilterChip(
                    selected = uiState.filtroCategoria == null,
                    onClick = { viewModel.onFiltroChange(null) },
                    label = { Text("Todos") }
                )
                FilterChip(
                    selected = uiState.filtroCategoria == CategoriaInsumo.INSUMO,
                    onClick = { viewModel.onFiltroChange(CategoriaInsumo.INSUMO) },
                    label = { Text("Insumos") }
                )
                FilterChip(
                    selected = uiState.filtroCategoria == CategoriaInsumo.EMBALAGEM,
                    onClick = { viewModel.onFiltroChange(CategoriaInsumo.EMBALAGEM) },
                    label = { Text("Embalagens") }
                )
            }

            when {
                uiState.isLoading -> Box(
                    modifier = Modifier.fillMaxSize(),
                    contentAlignment = Alignment.Center
                ) { CircularProgressIndicator(color = PurplePrimary) }

                uiState.insumosFiltrados.isEmpty() -> EmptyState(
                    icon = Icons.Outlined.Inventory2,
                    title = "Nenhum insumo cadastrado",
                    subtitle = "Cadastre insumos e embalagens com o custo por unidade para montar as fichas técnicas"
                )

                else -> LazyColumn(
                    modifier = Modifier.fillMaxSize(),
                    contentPadding = PaddingValues(
                        start = 16.dp, top = 4.dp, end = 16.dp, bottom = 88.dp
                    ),
                    verticalArrangement = Arrangement.spacedBy(8.dp)
                ) {
                    items(uiState.insumosFiltrados, key = { it.id }) { insumo ->
                        InsumoRow(
                            insumo = insumo,
                            onClick = { viewModel.onShowForm(insumo) },
                            onDelete = { viewModel.onShowDeleteConfirm(insumo) }
                        )
                    }
                }
            }
        }
    }

    if (uiState.showFormDialog) {
        InsumoFormDialog(
            insumo = uiState.insumoEmEdicao,
            onDismiss = viewModel::onHideForm,
            onConfirm = viewModel::save
        )
    }

    uiState.insumoParaExcluir?.let { insumo ->
        AlertDialog(
            onDismissRequest = viewModel::onHideDeleteConfirm,
            title = { Text("Excluir insumo?") },
            text = { Text("\"${insumo.nome}\" e seu histórico de preços serão excluídos.") },
            confirmButton = {
                TextButton(onClick = viewModel::delete) {
                    Text("Excluir", color = ErrorRed)
                }
            },
            dismissButton = {
                TextButton(onClick = viewModel::onHideDeleteConfirm) { Text("Cancelar") }
            }
        )
    }
}

@Composable
private fun InsumoRow(
    insumo: Insumo,
    onClick: () -> Unit,
    onDelete: () -> Unit
) {
    ElevatedCard(
        onClick = onClick,
        modifier = Modifier.fillMaxWidth(),
        elevation = CardDefaults.elevatedCardElevation(defaultElevation = 1.dp)
    ) {
        Row(
            modifier = Modifier.padding(horizontal = 16.dp, vertical = 12.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Column(modifier = Modifier.weight(1f)) {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Text(
                        text = insumo.nome,
                        style = MaterialTheme.typography.titleSmall,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                        modifier = Modifier.weight(1f, fill = false)
                    )
                    if (insumo.categoria == CategoriaInsumo.EMBALAGEM) {
                        Spacer(modifier = Modifier.width(8.dp))
                        Surface(shape = RoundedCornerShape(50), color = YellowContainer) {
                            Text(
                                text = "embalagem",
                                style = MaterialTheme.typography.labelSmall,
                                color = YellowOnContainer,
                                modifier = Modifier.padding(horizontal = 8.dp, vertical = 2.dp)
                            )
                        }
                    }
                }
                Text(
                    text = "1 ${insumo.unidadeCompra} = ${formatarFator(insumo.fatorConversao)} ${insumo.unidadeUso}",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }
            Column(horizontalAlignment = Alignment.End) {
                Text(
                    text = "R$ ${formatarMoeda(insumo.custoAtual)}/${insumo.unidadeCompra}",
                    style = MaterialTheme.typography.titleSmall,
                    fontWeight = FontWeight.SemiBold
                )
                Text(
                    text = "R$ ${formatarMoeda(insumo.custoPorUnidadeUso)}/${insumo.unidadeUso}",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }
            IconButton(onClick = onDelete) {
                Icon(
                    Icons.Default.Delete,
                    contentDescription = "Excluir insumo",
                    tint = MaterialTheme.colorScheme.onSurfaceVariant,
                    modifier = Modifier.size(20.dp)
                )
            }
        }
    }
}

private fun formatarFator(valor: Double): String =
    if (valor % 1.0 == 0.0) valor.toInt().toString() else String.format("%.2f", valor)

File: lib/presentation/insumos/insumos_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../domain/model/categoria_insumo.dart';
import '../../domain/model/insumo.dart';
import '../../theme/app_theme.dart';
import '../components/empty_state.dart';
import '../components/formatters.dart';
import 'insumo_form_dialog.dart';
import 'insumos_controller.dart';

class InsumosScreen extends ConsumerWidget {
  const InsumosScreen({super.key});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final insumosAsync = ref.watch(insumosStreamProvider);
    final uiState = ref.watch(insumosControllerProvider);
    final controller = ref.read(insumosControllerProvider.notifier);

    ref.listen(insumosControllerProvider.select((s) => s.error), (prev, next) {
      if (next != null) {
        ScaffoldMessenger.of(context).showSnackBar(SnackBar(content: Text(next)));
        controller.clearError();
      }
    });

    ref.listen(insumosControllerProvider.select((s) => s.showFormDialog), (prev, next) {
      if (next) {
        showDialog<void>(
          context: context,
          builder: (_) => InsumoFormDialog(
            insumo: ref.read(insumosControllerProvider).insumoEmEdicao,
            onConfirm: ({
              required nome,
              required categoria,
              required unidadeCompra,
              required unidadeUso,
              required fatorConversao,
              required custoAtual,
            }) =>
                controller.save(
              nome: nome,
              categoria: categoria,
              unidadeCompra: unidadeCompra,
              unidadeUso: unidadeUso,
              fatorConversao: fatorConversao,
              custoAtual: custoAtual,
            ),
          ),
        ).then((_) => controller.onHideForm());
      }
    });

    ref.listen(insumosControllerProvider.select((s) => s.insumoParaExcluir), (prev, next) {
      if (next != null) {
        showDialog<void>(
          context: context,
          builder: (_) => AlertDialog(
            title: const Text('Excluir insumo?'),
            content: Text('"${next.nome}" e seu histórico de preços serão excluídos.'),
            actions: [
              TextButton(
                onPressed: () => Navigator.of(context).pop(),
                child: const Text('Cancelar'),
              ),
              TextButton(
                onPressed: () {
                  Navigator.of(context).pop();
                  controller.delete();
                },
                child: const Text('Excluir', style: TextStyle(color: errorRed)),
              ),
            ],
          ),
        ).then((_) => controller.onHideDeleteConfirm());
      }
    });

    return Scaffold(
      appBar: AppBar(title: const Text('Biblioteca de Insumos')),
      floatingActionButton: FloatingActionButton.extended(
        onPressed: () => controller.onShowForm(),
        backgroundColor: yellowSecondary,
        foregroundColor: Colors.white,
        icon: const Icon(Icons.add),
        label: const Text('Novo Insumo'),
      ),
      body: Column(
        children: [
          Padding(
            padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 4),
            child: Row(
              children: [
                FilterChip(
                  label: const Text('Todos'),
                  selected: uiState.filtroCategoria == null,
                  onSelected: (_) => controller.onFiltroChange(null),
                ),
                const SizedBox(width: 8),
                FilterChip(
                  label: const Text('Insumos'),
                  selected: uiState.filtroCategoria == CategoriaInsumo.insumo,
                  onSelected: (_) => controller.onFiltroChange(CategoriaInsumo.insumo),
                ),
                const SizedBox(width: 8),
                FilterChip(
                  label: const Text('Embalagens'),
                  selected: uiState.filtroCategoria == CategoriaInsumo.embalagem,
                  onSelected: (_) =>
                      controller.onFiltroChange(CategoriaInsumo.embalagem),
                ),
              ],
            ),
          ),
          Expanded(
            child: insumosAsync.when(
              loading: () =>
                  const Center(child: CircularProgressIndicator(color: purplePrimary)),
              error: (e, _) => Center(child: Text('Erro: $e')),
              data: (insumos) {
                final filtrados = uiState.filtroCategoria == null
                    ? insumos
                    : insumos.where((i) => i.categoria == uiState.filtroCategoria).toList();
                if (filtrados.isEmpty) {
                  return const EmptyState(
                    icon: Icons.inventory_2_outlined,
                    title: 'Nenhum insumo cadastrado',
                    subtitle:
                        'Cadastre insumos e embalagens com o custo por unidade para montar as fichas técnicas',
                  );
                }
                return ListView.separated(
                  padding: const EdgeInsets.fromLTRB(16, 4, 16, 88),
                  itemCount: filtrados.length,
                  separatorBuilder: (_, _) => const SizedBox(height: 8),
                  itemBuilder: (context, i) {
                    final insumo = filtrados[i];
                    return _InsumoRow(
                      insumo: insumo,
                      onTap: () => controller.onShowForm(insumo),
                      onDelete: () => controller.onShowDeleteConfirm(insumo),
                    );
                  },
                );
              },
            ),
          ),
        ],
      ),
      bottomSheet: null,
    );
  }
}

class _InsumoRow extends StatelessWidget {
  const _InsumoRow({required this.insumo, required this.onTap, required this.onDelete});
  final Insumo insumo;
  final VoidCallback onTap;
  final VoidCallback onDelete;

  static String _formatarFator(double valor) {
    return valor % 1.0 == 0.0 ? valor.toInt().toString() : valor.toStringAsFixed(2);
  }

  @override
  Widget build(BuildContext context) {
    return Card(
      child: InkWell(
        onTap: onTap,
        child: Padding(
          padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 12),
          child: Row(
            children: [
              Expanded(
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Row(
                      children: [
                        Flexible(
                          child: Text(
                            insumo.nome,
                            maxLines: 1,
                            overflow: TextOverflow.ellipsis,
                            style: Theme.of(context).textTheme.titleSmall,
                          ),
                        ),
                        if (insumo.categoria == CategoriaInsumo.embalagem) ...[
                          const SizedBox(width: 8),
                          Container(
                            padding:
                                const EdgeInsets.symmetric(horizontal: 8, vertical: 2),
                            decoration: BoxDecoration(
                              color: yellowContainer,
                              borderRadius: BorderRadius.circular(50),
                            ),
                            child: Text(
                              'embalagem',
                              style: Theme.of(context)
                                  .textTheme
                                  .labelSmall
                                  ?.copyWith(color: yellowOnContainer),
                            ),
                          ),
                        ],
                      ],
                    ),
                    Text(
                      '1 ${insumo.unidadeCompra} = ${_formatarFator(insumo.fatorConversao)} ${insumo.unidadeUso}',
                      style: Theme.of(context).textTheme.bodySmall?.copyWith(
                            color: Theme.of(context).colorScheme.onSurfaceVariant,
                          ),
                    ),
                  ],
                ),
              ),
              Column(
                crossAxisAlignment: CrossAxisAlignment.end,
                children: [
                  Text(
                    'R\$ ${formatarMoeda(insumo.custoAtual)}/${insumo.unidadeCompra}',
                    style: Theme.of(context)
                        .textTheme
                        .titleSmall
                        ?.copyWith(fontWeight: FontWeight.w600),
                  ),
                  Text(
                    'R\$ ${formatarMoeda(insumo.custoPorUnidadeUso)}/${insumo.unidadeUso}',
                    style: Theme.of(context).textTheme.bodySmall?.copyWith(
                          color: Theme.of(context).colorScheme.onSurfaceVariant,
                        ),
                  ),
                ],
              ),
              IconButton(
                onPressed: onDelete,
                icon: Icon(
                  Icons.delete,
                  size: 20,
                  color: Theme.of(context).colorScheme.onSurfaceVariant,
                ),
                tooltip: 'Excluir insumo',
              ),
            ],
          ),
        ),
      ),
    );
  }
}
